src/net_processing.cpp

```diff
@@ -1507,6 +1517,18 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
 
         LOCK(cs_main);
 
+        // We might have announced the currently-being-connected tip using a
+        // compact block, which resulted in the peer sending a getblocks
+        // request, which we would otherwise respond to without the new block.
+        // To avoid this situation we simply verify that we are on our best
+        // known chain now. This is super overkill, but we handle it better
+        // for getheaders requests, and there are no known nodes which support
+        // compact blocks but still use getblocks to request blocks.
+        {
+            CValidationState dummy;
+            ActivateBestChain(dummy, Params());
```

Comment:
Mentioned on IRC as well:

An example of an unintended side-effect of this:
- miner mines a block, hands it to ProcessNewBlock()
- ProcessMessages thread simultaneously receives a GETBLOCKS
- ProcessMessages calls ActivateBestChain _just_ after miner finishes AcceptBlock() and releases cs_main.
- ConnectTip ends up getting called with a null block, activating the chain with the miner's new block, but forcing it to be read from disk
- Miner's ActivateBestChain is a no-op

The above is unlikely and harmless (minus the msec it takes to read/de-serialize from disk), but it's an example of how someone hammering a miner with GETBLOCKS would be capable of affecting validation.
Presumably with multi-threaded message processing, a node hammering with GETBLOCKS might be able to force all block reads to disk this way.

If the intention is simply to ensure that we've fully validated all known blocks at this moment, I'd be much more comfortable with some function that simply blocks until any current ActivateBestChain() completes. That would also protect us from oopses in the future when processing becomes multi-threaded.